Emby.Server.Implementations/SyncPlay/SyncPlayManager.cs

```diff
@@ -41,338 +44,396 @@ public class SyncPlayManager : ISyncPlayManager, IDisposable
+            JoinGroup(session, request, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Checks if a given session has joined a group.
+        /// </summary>
+        /// <param name="session">The session.</param>
+        /// <returns><c>true</c> if the session has joined a group, <c>false</c> otherwise.</returns>
+        private bool IsSessionInGroup(SessionInfo session)
+        {
+            lock (_mapsLock)
             {
-                return _groups.Values.Where(
-                    group => group.GetPlayingItemId().Equals(filterItemId) && HasAccessToItem(user, group.GetPlayingItemId())).Select(
-                    group => group.GetInfo()).ToList();
+                return _sessionToGroupMap.ContainsKey(session.Id);
             }
-            else
+        }
+
+        /// <summary>
+        /// Gets the group joined by the given session, if any.
+        /// </summary>
+        /// <param name="session">The session.</param>
+        /// <returns>The group.</returns>
+        private IGroupController FindJoinedGroup(SessionInfo session)
+        {
+            lock (_mapsLock)
             {
-                // Otherwise show all available groups
-                return _groups.Values.Where(
-                    group => HasAccessToItem(user, group.GetPlayingItemId())).Select(
-                    group => group.GetInfo()).ToList();
+                _sessionToGroupMap.TryGetValue(session.Id, out var group);
+                return group;
             }
         }
 
-        /// <inheritdoc />
-        public void HandleRequest(SessionInfo session, PlaybackRequest request, CancellationToken cancellationToken)
+        /// <summary>
+        /// Gets the group identifier joined by the given session, if any.
+        /// </summary>
+        /// <param name="session">The session.</param>
+        /// <returns>The group identifier if the session has joined a group, an empty identifier otherwise.</returns>
+        private Guid FindJoinedGroupId(SessionInfo session)
         {
-            var user = _userManager.GetUserById(session.UserId);
+            return FindJoinedGroup(session)?.GroupId ?? Guid.Empty;
+        }
 
-            if (user.SyncPlayAccess == SyncPlayAccess.None)
+        /// <summary>
+        /// Maps a session to a group.
+        /// </summary>
+        /// <param name="session">The session.</param>
+        /// <param name="group">The group.</param>
+        /// <exception cref="InvalidOperationException">Thrown when the user is in another group already.</exception>
+        private void AddSessionToGroup(SessionInfo session, IGroupController group)
+        {
+            if (session == null)
             {
-                _logger.LogWarning("HandleRequest: {0} does not have access to SyncPlay.", session.Id);
+                throw new InvalidOperationException("Session is null!");
+            }
 
-                var error = new GroupUpdate<string>()
+            lock (_mapsLock)
```

Comment:
`AddSessionToGroup` is a private function that is always called when `_mapsLock` is already held.

Reply:
Initially I put a remark in the description about locking but then I considered safer the addition of the lock in here as well, in case future devs missed it. I guess I should give more trust to whomever will make changes in here.